### ISSUE TITLE
Use SMAAEffect for 2 different scenes

### ISSUE THREAD

**Issue by bendxr** [vanruesc/postprocessing#158]
Hi, 

I have an issue having 2 scenes rendered with effects :
- sceneA : lots of meshes, several effects : smaa, outline, bloom, godrays, looked at with cameraA, responds to user interactions
- sceneB : 1 mesh, only 1 effect : smaa, looked at with cameraB, no user interaction, must not be affected by the effects of sceneA

```
// create the passes
renderPassA = new RenderPass(sceneA, cameraA);
renderPassB = new RenderPass(sceneB, cameraB);
effectsPassA = new EffectsPass(cameraA, smaa, outline, bloom, godrays);
effectsPassB = new EffectsPass(cameraB, smaa);

// the following works, but sceneB has no AA
renderPassA.clear = true;
renderPassA.renderToScreen = false;
effectsPassA.renderToScreen = true;
renderPassB.clear = false;
renderPassB.renderToScreen = true;
composer.addPass(renderPassA);
composer.addPass(effectsPassA);
composer.addPass(renderPassB);`


// the following does not work : sceneA is rendered with all effects
// sceneB is almost entirely hidden (only few vertices emerge from time to time) but it has got AA
renderPassA.clear = true;
renderPassA.renderToScreen = false;
effectsPassA.renderToScreen = false;
renderPassB.clear = false;
renderPassB.renderToScreen = false;
effectsPassB.renderToScreen = true;
composer.addPass(renderPassA);
composer.addPass(effectsPassA);
composer.addPass(renderPassB);
composer.addPass(effectsPassB);
```

What would you advise me to make it work ?

Best regards,

**Comment by vanruesc** [comment]
Hi,

the tricky part with this compositing problem is that the depth information is lost after the `EffectPass` has rendered its results.

Applying SMAA twice can produce artifacts around the silhouette of the object in `sceneB`. It's best to apply it to the entire image before other effects are applied.

For this use case, I suggest using a `MaskPass`. To keep the depth information intact for the `MaskPass`, we need a small custom pass that toggles the stencil test:

<details><summary>ToggleMaskPass (Custom Pass)</summary>
<p>

```js
import { Pass } from "postprocessing";

export class ToggleMaskPass extends Pass {

	/**
	 * Constructs a new toggle mask pass.
	 *
	 * @param {Boolean} [stencilTest=true] - The initial stencil test state.
	 */

	constructor(stencilTest = true) {

		super("ToggleMaskPass", null, null);

		this.needsSwap = false;

		/**
		 * Keeps track of the stencil test state.
		 *
		 * @private
		 * @type {Boolean}
		 */

		this.stencilTest = stencilTest;

	}

	/**
	 * Toggles the global stencil test.
	 *
	 * @param {WebGLRenderer} renderer - The renderer.
	 * @param {WebGLRenderTarget} inputBuffer - A frame buffer that contains the result of the previous pass.
	 * @param {WebGLRenderTarget} outputBuffer - A frame buffer that serves as the output render target unless this pass renders to screen.
	 * @param {Number} [deltaTime] - The time between the last frame and the current one in seconds.
	 * @param {Boolean} [stencilTest] - Indicates whether a stencil mask is active.
	 */

	render(renderer, inputBuffer, outputBuffer, deltaTime, stencilTest) {

		const stencil = renderer.state.buffers.stencil;

		stencil.setLocked(false);
		this.stencilTest = !this.stencilTest;
		stencil.setTest(this.stencilTest);
		stencil.setLocked(true);

	}

}
```

</p>
</details>

Try the following:

1. Set the `stencilBuffer` constructor parameter of your `EffectComposer` to `true`.
2. Render your entire scene (or `sceneA` followed by `sceneB`).
3. Render `sceneB` again using a [MaskPass](https://vanruesc.github.io/postprocessing/public/docs/class/src/passes/MaskPass.js~MaskPass.html). This creates a stencil mask and enables it.
4. Use `ToggleMaskPass` to disable the stencil mask.
5. Apply SMAA to the full scene.
6. Use `ToggleMaskPass` again to enable the stencil mask.
7. Apply other effects.
8. Use [ClearMaskPass](https://vanruesc.github.io/postprocessing/public/docs/class/src/passes/ClearMaskPass.js~ClearMaskPass.html).

I haven't tested this approach, but I think it should work.

Let me know if you run into trouble implementing this pass chain or if the results are not satisfactory.

**Comment by vanruesc** [comment]
Alternatively, you could also apply SMAA last. This might actually be a better option since the `Maskpass` introduces aliasing artifacts.

The pass chain would then look like this:

1. Set the `stencilBuffer` to `true`.
2. Render your entire scene.
3. Render `sceneB` using a `MaskPass`.
4. Apply effects (without SMAA).
5. Use `ClearMaskPass`.
5. Apply SMAA.

**Comment by bendxr** [comment]
Hi, 

Thank you for your detailed and precise answers, it helped me a lot to test what you proposed !
Unfortunately I didn't succeed in having the wanted rendering.

So I tested both of your solutions :
- 1st solution: both scenes are rendered, without any effects.
- 2nd solution : the best I 've had  is to have sceneA and sceneB both rendered with full effects (SMAA included), but with rendering sceneA into a MaskPass (if rendering sceneB into MaskPass, sceneA loses all its effects, while sceneB still have all the effects).
Does the fact that sceneA has a background (CubeMap) may complicate the setup ?

Morever in each solution I encoutered a new rendering inconvenient : some meshes of sceneA that are mainly transparent are now hiding my mesh on sceneB, which is more noticeable than a lack of SMAA.
It may be possible that I miss something in testing what you proposed but your answer let me think that there is no straightforward solution to this particular rendering setup so I will give up that issue for the moment.

Many thanks again for your advices,

**Comment by vanruesc** [comment]
I forgot to ask: is `sceneB` supposed to be part of `sceneA` or are the two scenes completely unrelated. In other words, does `sceneB` overwrite `sceneA` or do they share depth?

> your answer let me think that there is no straightforward solution to this particular rendering setup

This is indeed an advanced compositing problem and the implementation depends on the requirements and details of your use case.

> Does the fact that sceneA has a background (CubeMap) may complicate the setup ?

That shouldn't be a problem.

**Comment by bendxr** [comment]
Hi, 

Yes, sceneA and sceneB are completely unrelated (so each scene has its own camera), but sceneB (SMAA only) is meant to be always on top of sceneA (full effects).

**Comment by vanruesc** [comment]
Sorry for the delay!

> sceneB (SMAA only) is meant to be always on top of sceneA (full effects)

In that case, your original approach wasn't too far off:

```js
renderPassA = new RenderPass(sceneA, cameraA);
renderPassB = new RenderPass(sceneB, cameraB);
effectPassA = new EffectPass(cameraA, outline, bloom, godrays);
effectPassB = new EffectPass(cameraB, smaa);

renderPassB.clear = false;
effectsPassB.renderToScreen = true;

composer.addPass(renderPassA);	// Renders the main scene.
composer.addPass(effectPassA);	// Applies effects, but no SMAA.
composer.addPass(renderPassB);	// Renders on top of everything.
composer.addPass(effectPassB);	// Applies SMAA and renders to screen.
```

SMAA won't be as effective with this setup though, since it's forced to run last.

If you want to apply SMAA early to both scenes, you'll need to first render `sceneB`, then apply SMAA and save the result to an RGBA render target. Afterwards, you'd render `sceneA` and apply the effects together with SMAA. And in that last effect pass, you'd also use a `TextureEffect` that applies the saved result from `sceneB` using alpha blending.

I might be able to whip up an example this weekend.

**Comment by bendxr** [comment]
Hi, 

Thanks for your answer !
I'll test the solution with `TextureEffect` as soon as I can !

Regards,

**Comment by vanruesc** [comment]
> SMAA won't be as effective with this setup though, since it's forced to run last.

I tried implementing both approaches from https://github.com/vanruesc/postprocessing/issues/158#issuecomment-553655326 and found that the first one is better after all. Here are some examples:

- First approach: https://codesandbox.io/s/postprocessing-gb0t0
- Seconds approach: https://codesandbox.io/s/postprocessing-gppgh

Using a `TextureEffect` with alpha blending works, but there are some downsides:
- There are artifacts around the silhouette of the object in `sceneB`.
- It's necessary to enable `alpha` when creating the renderer to render geometry to an RGBA buffer.
- Applying SMAA twice is not great for performance.
- The setup is more complicated (requires a custom pass).

In conclusion, I'd say the first approach is better in terms of performance and results.

**Comment by bendxr** [comment]
Hi, 

Many thanks for your answer and examples.

I implemented first approach, but unfortunately I still have issues on the rendering of sceneB (the simplest), whereas sceneA is correctly renderered with all effects and SMAA.

My sceneB is at the moment only a textured cube, but only the corners are showing (cf picture) when trying to have SMAA on both scenes. Since your example is working, it's surely an issue on my side, but I didn't catched what is happening.

![sceneB_rendering_first_approach](https://user-images.githubusercontent.com/42646601/69434498-0fa7a580-0d3e-11ea-8515-75dc81275049.png)

best regards,

**Comment by vanruesc** [comment]
Closing this issue for now. Feel free to reopen if you need more help!
